react native local evaluation (#29)
amplitude/amplitude-dev-center: docs/experiment/sdks/react-native-sdk.md

!!!tip "Quick Start"

>     The right way to initialize the Experiment SDK depends on whether you use an

Reviewer: ```suggestion
    The correct way to initialize the Experiment SDK depends on whether you use an
```

>     Amplitude SDK for analytics or a third party (e.g. Segment).

Reviewer: ```suggestion
    Amplitude SDK for analytics or a third party tool like Segment.
```
Not sure why this is wrapped onto a new line.

    === "Amplitude Analytics"

        1. [Initialize the experiment client](#initialize)
        2. [Start the SDK](#start)
        3. [Access a flag's variant](#variant)

        ```js
        import { Experiment } from '@amplitude/experiment-react-native-client';

        // (1) Initialize the experiment client with Amplitude Analytics.
        const experiment = Experiment.initializeWithAmplitudeAnalytics(
            'DEPLOYMENT_KEY'
        );

        // (2) Start the SDK and await the promise result.
        await experiment.start();

        // (3) Lookup a flag's variant.
        const variant = experiment.variant('FLAG_KEY');
        if (variant.value === 'on') {
            // Flag is on
        } else {
            // Flag is off
        }
        ```

    === "Third Party Analytics"

        4. [Initialize the experiment client](#initialize)

>         5. [Start the SDK with the user](#fetch)

Author: ```suggestion
        5. [Start the SDK with the user](#start)
```

The initializer returns a singleton instance, so subsequent initializations for the same instance name will always return the initial instance. To create multiple instances, use the `instanceName` [configuration](#configuration).

=== "Amplitude Analytics"

    ```js
    import { Experiment } from '@amplitude/experiment-js-client';

    const experiment = initializeWithAmplitudeAnalytics('DEPLOYMENT_KEY');
    ```

    !!!note "Instance name"
        If you're using a custom instance name for analytics, you need to set the same value in the `instanceName` [configuration option](#configuration) in the experiment SDK, or visa versa.

=== "Third Party Analytics"

    ```js
    import { Experiment } from '@amplitude/experiment-js-client';

    const experiment = Experiment.initialize(
        'DEPLOYMENT_KEY',
        {
            exposureTrackingProvider: {
                track: (exposure) => {
                    // TODO: Implement exposure tracking
                    // analytics.track('$exposure', exposure)
                }
            }
        }
    );
    ```

#### Configuration

> The SDK client can be configured once on initialization.

Reviewer: ```suggestion
SDK client configuration occurs during initialization.
```

    | <div class="big-column">Name</div> | Description | Default Value |
    | --- | --- | --- |
    | `debug` | Enable additional debug logging within the SDK. Should be set to false in production builds. | `false` |
    | `fallbackVariant` | The default variant to fall back if a variant for the provided key doesn't exist. | `{}` |
    | `initialVariants` | An initial set of variants to access. This field is valuable for bootstrapping the client SDK with values rendered by the server using server-side rendering (SSR). | `{}` |
    | `serverZone` | Select the Amplitude data center to get flags and variants from, `us` or `eu`. | `us` |
    | `serverUrl` | The host to fetch remote evaluation variants from. For hitting the EU data center, use `serverZone`. | `https://api.lab.amplitude.com` |
    | `flagsServerUrl` | The host to fetch local evaluation flags from. For hitting the EU data center, use `serverZone`. | `https://flag.lab.amplitude.com` |
    | `fetchTimeoutMillis` | The timeout for fetching variants in milliseconds. | `10000` |
    | `retryFetchOnFailure` | Whether to retry variant fetches in the background if the request doesn't succeed. | `true` |

>     | `automaticExposureTracking` | If true, calling [`variant()`](#variant) will track an exposure event through the configured `exposureTrackingProvider`. If no exposure tracking provider is set, this configuration option does nothing.  | `true` |

Reviewer: ```suggestion
    | `automaticExposureTracking` | If true, calling [`variant()`](#variant) tracks an exposure event through the configured `exposureTrackingProvider`. If no exposure tracking provider is set, this configuration option does nothing.  | `true` |
```

    | `fetchOnStart` | If true, always [fetch](#fetch) remote evaluation variants on [start](#start). If false, never fetch on start. If undefined, dynamically determine whether to fetch on start. | `undefined` |
    | `pollOnStart` | Poll for local evaluation flag configuration updates once per minute on [start](#start). | `true` |

>     | `automaticFetchOnAmplitudeIdentityChange` | Only matters if you use the `initializeWithAmplitudeAnalytics` initialization function to seamlessly integrate with the Amplitude Analytics SDK. If `true` any change to the user ID, device ID or user properties from analytics will trigger the experiment SDK to fetch variants and update it's cache. | `false` |

Reviewer: ```suggestion
    | `automaticFetchOnAmplitudeIdentityChange` | Only matters if you use the `initializeWithAmplitudeAnalytics` initialization function to seamlessly integrate with the Amplitude Analytics SDK. If `true` any change to the user ID, device ID or user properties from analytics triggers the experiment SDK to fetch variants and update it's cache. | `false` |
```

    | `userProvider` | An interface used to provide the user object to `fetch()` when called. See [Experiment User](https://developers.experiment.amplitude.com/docs/experiment-user#user-providers) for more information. | `null` |
    | `exposureTrackingProvider` | Implement and configure this interface to track exposure events through the experiment SDK, either automatically or explicitly. | `null` |
    | `instanceName` | Custom instance name for experiment SDK instance. **The value of this field is case-sensitive.** | `null` |

!!!info "EU Data Center"
    If you're using Amplitude's EU data center, configure the `serverZone` option on initialization to `eu`.

#### Integrations

> If you use either Amplitude or Segment Analytics SDKs to track events into Amplitude, you'll want to set up an integration on initialization. Integrations automatically implement [provider](#providers) interfaces to enable a more streamlined developer experience by making it easier to **manage user identity** and **track exposures events**.

Reviewer: ```suggestion
If you use either Amplitude or Segment Analytics SDKs to track events into Amplitude, Amplitude recommends that you set up an integration on initialization. Integrations implement [provider](#providers) interfaces to enable a more streamlined developer experience by making it easier to **manage user identity** and **track exposures events**.
```

    Using the integration initializer will automatically configure implementations of the [user provider](#user-provider) and [exposure tracking provider](#exposure-tracking-provider) interfaces to pull user data from the Amplitude Analytics SDK and track exposure events.

???segment "Segment integration (click to open)"

    Experiment's integration with Segment Analytics must be configured manually. The Experiment SDK must then be configured on initialization with an instance of the the exposure tracking provider. Make sure this happens _after_ the analytics SDK has been loaded an initialized.

    ```js
    analytics.ready(() => {
        const experiment =  Experiment.initialize('DEPLOYMENT_KEY', {
            exposureTrackingProvider: {
                track: (exposure) => {
                    analytics.track('$exposure', exposure)
                }
            }
        });
    });
    ```

    When [starting the SDK](#start), pass the segment anonymous ID and user ID for the device ID and user ID, respectively.

    ```js
    await experiment.start({
        user_id: analytics.user().id(),
        device_id: analytics.user().analyticsId(),
    });
    ```

### Start

> Start the SDK by getting flag configurations from the server and fetching remote evaluation variants for the user. The SDK is ready once the returned promise resolves.

Reviewer: ```suggestion
The Experiment SDK starts when you get flag configurations from the server and fetch remote evaluation variants for the user. The SDK is ready after the returned promise resolves.
```

Author: ```suggestion
Start the Experiment SDK to get flag configurations from the server and fetch remote evaluation variants for the user. The SDK is ready once the returned promise resolves.
```

```js
start(user?: ExperimentUser): Promise<void>
```

| Parameter | Requirement | Description |
| --- | --- | --- |

> | `user` | optional | Explicit [user](../general/data-model.md#users) information to pass with the request to fetch variants. This user information is merged with user information provided from [integrations](#integrations) via the [user provider](#user-provider), preferring properties passed explicitly to `fetch()` over provided properties. Also sets the user in the SDK for reuse. | `undefined` |

Reviewer: ```suggestion
| `user` | optional | Explicit [user](../general/data-model.md#users) information to pass with the request to fetch variants. This user information merges with user information from any [integrations](#integrations) through the [user provider](#user-provider), and prefers properties passed explicitly to `fetch()` over provided properties. Also sets the user in the SDK for reuse. | `undefined` |
```

Call `start()` when your application is initializing, after user information is available to use to evaluate or [fetch](#fetch) variants. The returned promise resolves after loading local evaluation flag configurations and fetching remote evaluation variants.

> Configure the behavior of `start()` by setting `fetchOnStart` in the SDK configuration on initialization to **improve performance based on the needs of your application**.

Reviewer: ```suggestion
Set `fetchOnStart` in the SDK configuration to set the behavior of `start()` to improve the performance of your application.
```

* If your application always relies on remote evaluation on startup, set `fetchOnStart` to `true` to load flag configurations and fetch remote evaluation variants simultaneously.
* If your application never relies on remote evaluation, set `fetchOnStart` to `false` to avoid increased startup latency if someone accidentally creates a remote evaluation flag.

> * If your application relies on remote evaluation, but not right at startup, you may set `fetchOnStart` to `false` and call `fetch()` and await the promise separately.

Reviewer: ```suggestion
* If your application relies on remote evaluation, but not right at startup, set `fetchOnStart` to `false` and call `fetch()` and await the promise separately.
```
